File: cpp/trees/LevelOrder.cpp
/*
Q4. Perform level order traversal (BFS) of a binary tree.
*/

#include <bits/stdc++.h>
using namespace std;

struct Node {
    int val;
    Node *left, *right;
    Node(int v): val(v), left(NULL), right(NULL) {}
};

class Solution {
public:
    void levelOrder(Node* root) {
        if (!root) return;
        queue<Node*> q;
        q.push(root);
        while (!q.empty()) {
            int size = q.size();
            while (size--) {
                Node* cur = q.front(); q.pop();
                cout << cur->val << " ";
                if (cur->left) q.push(cur->left);
                if (cur->right) q.push(cur->right);
            }
            cout << endl;
        }
    }

    Node* buildTree() {
        Node* root = new Node(1);
        root->left = new Node(2);
        root->right = new Node(3);
        root->left->left = new Node(4);
        root->left->right = new Node(5);
        root->right->right = new Node(6);
        return root;
    }

    void solve() {
        Node* root = buildTree();
        cout << "Level Order Traversal:\n";
        levelOrder(root);
    }
};

int main() {
    Solution sol;
    sol.solve();
}

File: python/trees/LevelOrder.py
"""
Q4. Perform level order traversal (BFS) of a binary tree.
"""

from collections import deque

class Node:
    def __init__(self, val):
        self.val = val
        self.left = self.right = None

class Solution:
    def levelOrder(self, root):
        if not root:
            return
        q = deque([root])
        while q:
            for _ in range(len(q)):
                node = q.popleft()
                print(node.val, end=" ")
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            print()

    def buildTree(self):
        root = Node(1)
        root.left = Node(2)
        root.right = Node(3)
        root.left.left = Node(4)
        root.left.right = Node(5)
        root.right.right = Node(6)
        return root

    def solve(self):
        root = self.buildTree()
        print("Level Order Traversal:")
        self.levelOrder(root)

if __name__ == "__main__":
    Solution().solve()
